src/gromacs/nblib/topology.cpp

```diff
@@ -63,14 +76,33 @@ t_blocka TopologyBuilder::createExclusionsList() const
         //! duplicate the exclusionBlockPerMolecule for the number of Molecules of (numMols)
         for (size_t i = 0; i < numMols; ++i)
         {
-            std::copy(std::begin(exclusionBlockPerMolecule), std::end(exclusionBlockPerMolecule),
-                      std::back_inserter(exclusionBlockGlobal));
+            auto shiftedExclusions = exclusionBlockPerMolecule;
+            //! shift atom numbers by atomNumberOffset
+            for (auto& localBlock : shiftedExclusions)
+            {
+                std::transform(std::begin(localBlock.atomNumber), std::end(localBlock.atomNumber),
+                               std::begin(localBlock.atomNumber),
+                               [atomNumberOffset](auto i){ return i + atomNumberOffset; });
+            }
+
+            std::copy(std::begin(shiftedExclusions), std::end(shiftedExclusions),
+                           std::back_inserter(exclusionBlockGlobal));
+
+            atomNumberOffset += molecule.numAtomsInMolecule();
         }
     }
 
+    size_t numberOfExclusions = std::accumulate(std::begin(exclusionBlockGlobal),
+                                                std::end(exclusionBlockGlobal), size_t(0),
+                                                [](size_t acc, const auto &block)
+                                                { return acc + block.atomNumber.size();});
+
     //! At the very end, convert the exclusionBlockGlobal into
     //! a massive t_blocka and return
     t_blocka tBlockGlobal;
+    snew(tBlockGlobal.index, numberOfExclusions + 1);
+    snew(tBlockGlobal.a, numberOfExclusions + 1);
```

Comment:
I was mistaken about the structure of t_blocka. It is actually one list with all the exclusions and one list with the indices of exclusion block start and ends:
blocks (.index): [0, endBlock1(==startBlock2 - 1), endBlock2(==startBlock3 - 1), ...]
exclusions (.a): [i, j, ..., IndexEndBlock1, IndexEndBlock1 + 1, k, l, ..., IndexEndBlock2, ...]
See src/topology/block.h for the definition.
For LJ particles index==a but in general index.size()==numAtoms + 1 and a.size()==numExclusions